Sprint 11

WayOfTheDev opened the pull request:
Спринт 11.

WayOfTheDev committed:
Устранение гонки

Спринт 11. Многопоточная среда и устранение гонки

WayOfTheDev committed:
Спринт 11. Запросы для экрана профиля

Добавлен функционал по получению базовой информации профиля.
Получение аватарки пользователя
Изменение работы со старого API на новый

WayOfTheDev committed:
Спринт 11. Использование Kingfisher

Добавление функционала загрузки аватара с помощью Kingfisher

WayOfTheDev committed:
Спринт 11. Keychain

Замена хранения данных, на безопасный способ с помощью Keychain

WayOfTheDev committed:
Спринт 11. Задание по 11 спринту.

Сверстан кодом SplashViewController и настроен UIWindow

Helvetii951 commented on Image Feed/OAuth2/ProfileImageService.swift:
```diff
@@ -0,0 +1,85 @@
+import UIKit
+
+// MARK: - UserResult
+struct UserResult: Codable {
+    let profileImage: ProfileImage?
+
+    struct ProfileImage: Codable {
+        let small: String?
+        let medium: String?
+        let large: String?
+    }
+
+    enum CodingKeys: String, CodingKey {
+        case profileImage = "profile_image"
+    }
+}
+
+// MARK: - ProfileImageService
+final class ProfileImageService {
+    
+    // MARK: - Singleton
+    static let shared = ProfileImageService()
```
Лучше создавать для каждой сущности один файл. Так навигация по проекту будет понятнее. Т.е. стоит создать для `UserResult` отдельный файл.

Helvetii951 commented on Image Feed/OAuth2/ProfileService.swift:
```diff
@@ -0,0 +1,82 @@
+import Foundation
+
+// MARK: - ProfileResult
+struct ProfileResult: Codable {
+    let username: String
+    let name: String?
+    let bio: String?
+    let profileImage: ProfileImage?
+
+    enum CodingKeys: String, CodingKey {
+        case username, name, bio
+        case profileImage = "profile_image"
+    }
+    
+    struct ProfileImage: Codable {
+        let small: String?
+        let medium: String?
+        let large: String?
+    }
+}
+
+// MARK: - Profile
+struct Profile {
+    let username: String
+    let name: String
+    let loginName: String
```
Для этих структур тоже стоит создать отдельные файлы.

Helvetii951 commented on Image Feed/SplashViewController/SplashViewController.swift:
```diff
@@ -5,9 +5,28 @@ final class SplashViewController: UIViewController {
+    private let profileService = ProfileService.shared
+    
+    // MARK: - UI Elements
+    private let splashImageView: UIImageView = {
+        let imageView = UIImageView(image: UIImage(named: "splash_screen_logo"))
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
```
Рекомендую всегда использовать [ленивые свойства](https://www.swiftbysundell.com/articles/using-lazy-properties-in-swift/)для объявления подобных вьюх.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.
```swift
private lazy var myAweasomeButton : UIButton = {
  let button = UIButton()
  button.titleLabel?.font = UIFont.systemFont(ofSize: 120, weight: .heavy)
  // настройка других свойств 
  // ...
  return button
}()
```

WayOfTheDev committed:
Спринт 11. Исправления после ревью

Вынес в отдельные файлы структуры и протоколы
Использовал ленивые свойства
Внес правки по рекомендациям

WayOfTheDev committed:
Спринт 11. Правки

Добавил инициализатор в ProfileImageService и ProfileService, теперь Синглтон :) (сори, я забыл про комментарии на ЯП, и только в гите посмотрел)
